### PR TITLE
Make discriminator read-only by default

### LINKED ISSUE #21154: Discriminator should be read-only

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #21154

### CODE REVIEW

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 1]
```diff
@@ -4112,7 +4112,8 @@ private InternalPropertyBuilder GetOrCreateDiscriminatorProperty(Type type, stri
                 type ?? discriminatorProperty?.ClrType ?? _defaultDiscriminatorType,
                 name ?? discriminatorProperty?.Name ?? _defaultDiscriminatorName,
                 typeConfigurationSource: type != null ? configurationSource : (ConfigurationSource?)null,
-                configurationSource: configurationSource);
+                configurationSource: configurationSource)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs:4115]:
```suggestion
                configurationSource)
```

File: test/EFCore.Relational.Tests/Migrations/Internal/MigrationsModelDifferTest.cs [hunk 2]
```diff
@@ -5712,11 +5677,10 @@ public void Change_TPH_to_TPT_with_FKs_and_seed_data()
                             operation.KeyValues,
                             v => Assert.Equal(22, v));
 
-                        Assert.Equal(new[] { "Discriminator", "PreyId" }, operation.Columns);
+                        Assert.Equal(new[] { "PreyId" }, operation.Columns);
                         Assert.Null(operation.ColumnTypes);
                         AssertMultidimensionalArray(
                             operation.Values,
-                            v => Assert.Equal("Dog", v),
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Tests/Migrations/Internal/MigrationsModelDifferTest.cs:5719]: This is incorrect. If the discriminator value is not updated the seed entities end up with a null there.

**Comment by ajcvickers** [comment]
@AndriySvyryd I've been thinking about the force/recreate options. First, for the force option, we would have to assume that the property is really not read-only, even though the metadata says it is. We can't tell if a discriminator was explicitly configured as read-only (without putting more information in the model.) This also requires us to build the mechanism to allow forcing a write in the update pipeline.

Second, if model data updates don't work correctly when a discriminator is marked read-only, then doesn't that mean that they may not work if any property is marked as read-only? Or are the errors here specific to it being a discriminator?

Do you think it's still correct to go the forced update route? Or if we have to fix read-only properties anyway, then should discriminators really be special-cased?

**Comment by AndriySvyryd** [comment]
> Second, if model data updates don't work correctly when a discriminator is marked read-only, then doesn't that mean that they may not work if any property is marked as read-only? Or are the errors here specific to it being a discriminator?

So far we've assumed that readonly properties would be store generated. Instead we can check and if they aren't then we'll recreate the row.

> Do you think it's still correct to go the forced update route? Or if we have to fix read-only properties anyway, then should discriminators really be special-cased?

I don't think it's correct, but it will be a better experience in most cases. However I don't feel strongly.

**Comment by AndriySvyryd** [comment]
I've convinced myself that recreating the row is fine as there's still the workaround of configuring the discriminator as writable just while creating the migrations

**Comment by ajcvickers** [comment]
@AndriySvyryd Would appreciate if you can take this over. Doesn't need to go into preview 8 at this point.

**Pull request merged**